lib/coach/middleware_validator.rb

```diff
@@ -21,14 +22,18 @@ def validated_provides!
 
     private
 
+    attr_reader :previous_middlewares
+
     def missing_requirements
-      @middleware.requirements - provided_by_chain
+      @missing_requirements ||= @middleware.requirements - provided_by_chain
```

Comment:
Don't think memoising adds very much here - set subtraction should be super cheap